command/agent/keyring_endpoint.go

```diff
@@ -4,12 +4,79 @@
 package agent
 
 import (
+	"crypto/ed25519"
+	"fmt"
 	"net/http"
 	"strings"
+	"time"
 
+	"github.com/go-jose/go-jose/v3"
+	"github.com/hashicorp/nomad/helper"
 	"github.com/hashicorp/nomad/nomad/structs"
 )
 
+// jwksMinMaxAge is the minimum amount of time the JWKS endpoint will instruct
+// consumers to cache a response for.
+const jwksMinMaxAge = 15 * time.Minute
+
+// JWKSRequest is used to handle JWKS requests. JWKS stands for JSON Web Key
+// Sets and returns the public keys used for signing workload identities. Third
+// parties may use this endpoint to validate workload identities. Consumers
+// should cache this endpoint, preferably until an unknown kid is encountered.
+func (s *HTTPServer) JWKSRequest(resp http.ResponseWriter, req *http.Request) (any, error) {
+	if req.Method != http.MethodGet {
+		return nil, CodedError(405, ErrInvalidMethod)
+	}
+
+	args := structs.GenericRequest{}
+	if s.parse(resp, req, &args.Region, &args.QueryOptions) {
+		return nil, nil
+	}
+
+	var rpcReply structs.KeyringListPublicResponse
+	if err := s.agent.RPC("Keyring.ListPublic", &args, &rpcReply); err != nil {
+		return nil, err
+	}
+	setMeta(resp, &rpcReply.QueryMeta)
+
+	// Key set will change after max(CreateTime) + RotationThreshold.
+	var newestKey int64
+	jwks := make([]jose.JSONWebKey, 0, len(rpcReply.PublicKeys))
+	for _, pubKey := range rpcReply.PublicKeys {
+		if pubKey.CreateTime > newestKey {
+			newestKey = pubKey.CreateTime
+		}
+
+		jwk := jose.JSONWebKey{
+			KeyID:     pubKey.KeyID,
+			Algorithm: pubKey.Algorithm,
+			Use:       pubKey.Use,
+		}
+		switch alg := pubKey.Algorithm; alg {
+		case structs.PubKeyAlgEdDSA:
+			// Convert public key bytes to an ed25519 public key
+			jwk.Key = ed25519.PublicKey(pubKey.PublicKey)
+		default:
+			s.logger.Warn("unknown public key algorithm. server is likely newer than client", "alg", alg)
+		}
```

Comment:
Yeah I struggled with what responsibilities to put in the RPC vs HTTP handlers. If I understand your proposal it would mean the `KeyringPublicKey` struct the RPC handler returns would get changed to something like:

```go
type KeyringPublicKey struct {
	KeyID     string `json:"kid"`

	// Always "OKP" (octet key pair) currently
	KeyType     string `json:"kty"`

	PublicKey []byte `json:"x"`

	// Always "EdDSA" currently
	Algorithm string `json:"crv"`

	// Always "Ed25519" currently, must match alg
	Curve string `json:"alg"`

	// Always "sig" currently
	Use string `json:"use"`

	// Purely for Nomad's cache control use. Not part of JWK spec.
	CreateTime int64 `json:"-"`
}
```

Then we would just JSON marshal `struct { Keys []KeyringPublicKey }` in the HTTP handler and not even touch `go-jose.JSONWebKey`.

#### Reason 1 Against: Too DIY

I avoided this initially because it felt too DIY. The nice thing about `go-jose.JSONWebKey` is that it knows how to convert key material to the appropriate algorithm (agl), curve (crv), and type (kty) fields. In the future if we add other key types, as long as we updated the mapping of `Algorithm -> Go Type`, go-jose would ensure it was marshalled properly.

That being said go-jose is "saving" us 2 fields currently (`kty` and `crv`)... and `crv` we're effectively setting ourselves by doing the type conversion (that as you know is easy to forget!).

So my "too DIY to skip go-jose" argument seems weak at best...

#### Reason 2 Against: Validating JWTs in the Agent

But there's another reason to do the conversion here: if we ever want to validate JWTs in the Agent, the Agent needs to do the conversion we're doing here. There's no way to avoid converting `Key []byte -> ed25519.PublicKey` at some point in order to use it.

The [auth middleware used for the Task API and the WhoAmI RPC](https://github.com/hashicorp/nomad/blob/v1.6.0/command/agent/http.go#L1122-L1181) would both benefit from being able to validate JWTs using these PublicKeys. In my other branch I have a `Keyring.ListPublic -> PublicKeyCache` conversion implemented so the JWKS endpoint and auth middleware could share the underlying public keys in their native Go types.

I think this is reason enough to keep `go-jose.JSONWebKey` around even though the type conversions are gross and error prone? I think we can keep the problematic conversion to 1 or 2 places in code.

### Alternative: JSONWebKeySet internally

**Plan:** just have the RPC return `go-jose.JSONWebKeySet` and just let the HTTP handler serialize it.

**Pro:** All of the key handling is the `encrypter` and `keyring_endpoint` files.

**Con:**

 I don't think it's safe to send [go-jose's `JSONWebKey` struct](https://pkg.go.dev/github.com/go-jose/go-jose/v3#JSONWebKey) over our msgpack RPC. Maybe it is, and I could test it, but even if it is today I'm worried a change in either go-jose, msgpack, or even Go's crypto library could break things.

My main concern is [`JSONWebKey.Key any`](https://pkg.go.dev/github.com/go-jose/go-jose/v3#JSONWebKey) type confusion: go-jose relies on the key's concrete type to properly emit the right JWKS JSON (through a [custom marshal implementation](https://github.com/go-jose/go-jose/blob/v3.0.0/jwk.go#L92) and [internal struct](https://github.com/go-jose/go-jose/blob/v3.0.0/jwk.go#L41-L68)). If the RPC/msgpack layer were to lose the concrete type information, [go-jose may guess that it's a symmetric encryption key and cause a really confusing error](https://github.com/go-jose/go-jose/blob/v3.0.0/jwk.go#L109-L110) (that I caused the other day and you helped me debug!).

I'm also a little worried that if we started using X509 certificates in the future that the [`x509.Certificate`](https://pkg.go.dev/crypto/x509#Certificate) might not roundtrip cleanly through msgpack, so we'd silently change things like timestamps?

I don't know... these may be unreasonable concerns, but `explicit is better than implicit` seems doubly true to me when it comes to cryptographic materials and `Key interface{}` is the definition of _implicit._

Reply:
>  If I understand your proposal it would mean the KeyringPublicKey struct the RPC handler returns would get changed to something like:
> ...

Yeah that's pretty much what I was thinking. To be clear, I'm less concerned about where _in the code_ the conversion is happening than avoiding weird cross-version problems in doing that conversion with info we got from the server. But...

> The auth middleware used for the Task API and the WhoAmI RPC would both benefit from being able to validate JWTs using these PublicKeys.

I think I'm probably sold on the basis of this alone. We know we're likely to want to do this in the near-ish future.